src/modules/ecdsa_adaptor/main_impl.h

```diff
@@ -0,0 +1,311 @@
+    secp256k1_scalar dleq_proof_s, dleq_proof_e;
+    secp256k1_scalar msg;
+    secp256k1_ge q;
+    secp256k1_ge r, rp;
+    secp256k1_scalar sp;
+    secp256k1_scalar sigr;
+    secp256k1_ge adaptor_ge;
+    secp256k1_ge rhs;
+    secp256k1_gej lhs;
+
+    VERIFY_CHECK(ctx != NULL);
+    ARG_CHECK(secp256k1_ecmult_context_is_built(&ctx->ecmult_ctx));
+    ARG_CHECK(adaptor_sig65 != NULL);
+    ARG_CHECK(pubkey != NULL);
+    ARG_CHECK(msg32 != NULL);
+    ARG_CHECK(adaptor != NULL);
+    ARG_CHECK(adaptor_proof97 != NULL);
+
+    /* 1. DLEQ_verify((G,R'),(Y, R)) */
+    if (!secp256k1_ecdsa_adaptor_proof_deserialize(&rp, &dleq_proof_s, &dleq_proof_e, adaptor_proof97)) {
+        return 0;
+    }
+    if (!secp256k1_ecdsa_adaptor_sig_deserialize(&r, &sigr, &sp, adaptor_sig65)) {
+        return 0;
+    }
+    if (!secp256k1_pubkey_load(ctx, &adaptor_ge, adaptor)) {
+        return 0;
+    }
+    if(!secp256k1_dleq_verify(&ctx->ecmult_ctx, (unsigned char *)"ECDSAAdaptorSig", &dleq_proof_s, &dleq_proof_e, &rp, &adaptor_ge, &r)) {
+        return 0;
+    }
+
+    /* 2. return x_coord(R') == x_coord(s'⁻¹(H(m) * G + x_coord(R) * X)) */
+    secp256k1_scalar_set_b32(&msg, msg32, NULL);
+    if (!secp256k1_pubkey_load(ctx, &q, pubkey)) {
+        return 0;
+    }
+    if (!secp256k1_ecdsa_adaptor_sig_verify_helper(ctx, &rhs, &sigr, &sp, &q, &msg)) {
+        return 0;
+    }
+
+    secp256k1_gej_set_ge(&lhs, &rp);
+    secp256k1_ge_neg(&rhs, &rhs);
+    secp256k1_gej_add_ge_var(&lhs, &lhs, &rhs, NULL);
+    return secp256k1_gej_is_infinity(&lhs);
+}
+
+int secp256k1_ecdsa_adaptor_adapt(const secp256k1_context* ctx, secp256k1_ecdsa_signature *sig, const unsigned char *adaptor_secret32, const unsigned char *adaptor_sig65) {
+    secp256k1_scalar adaptor_secret;
+    secp256k1_scalar sp;
+    secp256k1_scalar s;
+    secp256k1_scalar sigr;
+    int overflow;
+    unsigned char buf32[32];
+    int high;
+
+    VERIFY_CHECK(ctx != NULL);
+    ARG_CHECK(sig != NULL);
+    ARG_CHECK(adaptor_secret32 != NULL);
+    ARG_CHECK(adaptor_sig65 != NULL);
+
+    secp256k1_scalar_set_b32(&adaptor_secret, adaptor_secret32, &overflow);
+    if (overflow) {
+        return 0;
+    }
+
+    if (!secp256k1_ecdsa_adaptor_sig_deserialize(NULL, &sigr, &sp, adaptor_sig65)) {
+        secp256k1_scalar_clear(&adaptor_secret);
+        return 0;
+    }
+    secp256k1_scalar_inverse(&s, &adaptor_secret);
+    secp256k1_scalar_mul(&s, &s, &sp);
+    high = secp256k1_scalar_is_high(&s);
+    secp256k1_scalar_cond_negate(&s, high);
+
+    secp256k1_ecdsa_signature_save(sig, &sigr, &s);
+
+    memset(buf32, 0, sizeof(buf32));
+    secp256k1_scalar_clear(&adaptor_secret);
+    secp256k1_scalar_clear(&sp);
+    secp256k1_scalar_clear(&s);
+
+    return 1;
+}
+
+int secp256k1_ecdsa_adaptor_extract_secret(const secp256k1_context* ctx, unsigned char *adaptor_secret32, const secp256k1_ecdsa_signature *sig, const unsigned char *adaptor_sig65, const secp256k1_pubkey *adaptor) {
```

Comment:
What do you think about enforcing the logic that if the correct secret is extracted then this **always** implies the signature is correct (for both ECDSA and Schnorr). This is what I went for here: https://github.com/LLFourn/secp256kfun/blob/master/ecdsa_fun/src/adaptor/mod.rs#L256. I check that the R_x in the ciphertext is the same as the R_x in the final signature.

This is to protect against the weird but I *think* not unimaginable situation where an attacker sends you a `sig` that has a correct `s` value but a bogus `R` value off-chain. Since you successfully extract the secret the programmer may think that this implies the signature was valid too. If they try to use it later they may end up in a pickle.

This is not normally an issue from a protocol standpoint since it is usually the creator of the encrypted signature that is trying to extract the secret so the R value is not interesting. I'm not sure this is always the case though. I think there *is* no case where you want to extract the secret from an invalid signature.

It is also just faster to extract/verify at the same time rather than verify then extract.

Reply:
Good idea and very cheap to do. Added a fix that compares the R values. I haven't thought through all the edge cases to guarantee that the signature is correct.

Reply:
I think as long as you are not enforcing low_s in verification then making sure that the R_x values are the same is sufficient as long as the encrypted signature is valid and you successfully recover the correct key.